src/user_crate/mod.rs

```diff
@@ -210,12 +211,76 @@ fn parse_source_and_deps(code_and_deps: &str) -> eyre::Result<(syn::Block, toml:
     let user_dependencies: toml::value::Table =
         toml::from_str(&deps_block).map_err(PlRustError::ParsingDependenciesBlock)?;
 
+    if crate::gucs::allow_listed_dependencies_only() {
+        validate_dependences_are_allowed(&user_dependencies);
+    }
+
     let user_code: syn::Block =
         syn::parse_str(&code_block).map_err(PlRustError::ParsingCodeBlock)?;
 
     Ok((user_code, user_dependencies))
 }
 
+#[tracing::instrument(level = "debug", skip_all)]
+fn validate_dependences_are_allowed(user_dependencies: &toml::value::Table) {
+    let allowed_deps = crate::gucs::get_allow_listed_dependencies();
+    let mut unsupported_deps = std::collections::HashMap::<String, toml::value::Value>::new();
+
+    for (dep, val) in user_dependencies {
+        if !allowed_deps.contains_key(dep) {
+            unsupported_deps.insert(dep.to_string(), val.clone());
+            continue;
+        }
+
+        match val {
+            toml::Value::String(ver) => {
+                let req = semver::VersionReq::parse(ver.as_str()).unwrap();
+
+                // Check if the allowed dependency is of format String or toml::Table
+                // foo = "1.0.0" vs foo = { version = "1.0.0", features = ["foo", "bar"]}
+                match allowed_deps.get(dep).unwrap() {
+                    toml::Value::String(allowed_deps_ver) => {
+                        if !req.matches(&semver::Version::parse(allowed_deps_ver).unwrap()) {
+                            unsupported_deps.insert(dep.to_string(), val.clone());
+                        }
+                    }
+                    toml::Value::Table(allowed_deps_vals) => {
+                        if !allowed_deps_vals.contains_key("version") {
+                            pgx::error!(
+                                "The allowed dependencies file does not specify a version for the crate {}",
+                                dep
+                            );
+                        }
+
+                        if !req.matches(
+                            &semver::Version::parse(
+                                &allowed_deps_vals.get("version").unwrap().as_str().unwrap(),
+                            )
+                            .unwrap(),
+                        ) {
+                            unsupported_deps.insert(dep.to_string(), val.clone());
+                        }
+                    }
+                    _ => {
+                        pgx::error!("Currently only toml Inline Table and String format is supported for dependencies");
+                    }
+                }
+            }
+            _ => {
+                pgx::error!("The toml::Value {:?} is not currently supported", val);
```

Comment:
These kinds of bounds should not go unenforced just because the passlist is ℵ₁.

Reply:
I think I'm misunderstanding here.
IIUC this means that if the `toml::Value` doesn't match `String` or `Table`, we error out because we don't know how to interpret something like

```
[dependencies]
pi = 3.14
foo = true
```

Reply:
I mean control flow should not arrive here only when we plan to restrict the number of dependencies, or else submitting an unhandled form, and then restricting dependencies in a way that affects zero dependencies, causes an unexpected UX: the dependency list is now rejected.